035-ca-abstraction.md

```diff
@@ -0,0 +1,396 @@
+
+There are multiple pieces of information that need to be stored about each end entity:
+
+* Whether a (new) certificate has been requested
+* Whether a (new) certificate is waiting (for trust in the issuing CA's certificate)
+* The issued certificate chain
+* The issued key
+
+It is proposed that these be stored in a separate `Secrets` than the one from which the (eventually) usable certificate is consumed, using the pattern `${cluster}-issuing-${type}`.
+For example, 
+for certificates being issued to brokers we'd use the name `${cluster}-issuing-broker-certs`
+while the certificates consumed by brokers is named using the pattern `${cluster}-broker-certs`.
+This prevents consumers (e.g. via Secret volume mounts) from being able to observe certificates that are not yet trusted,
+and also the formats of these Secrets to evolve separately.
+A similar separation would be applied for other certificates (ZK, Kafka exporter, `KafkaUsers` and so one); each would have its own value for the `type` placeholder in the `metadata.name`.
+Note that these extra certificates need not exist outside times that certificates need to be renewed.
+
+The structure of these `issuing` certificates could be:
+
+```yaml
+metadata:
+  name: ${cluster}-issuing-${type}
+type: strimzi.io/issuing
+data:
+  ${id}.${state}: <data>
+```
+
+The `${id}` would allow, for example, a single Secret to be used for all broker certificates, mirroring the situation with the `${cluster}-broker-certs` Secret. 
+
+The `state` would be one of `requested`, `pending`, or `issued`.
+The `data` would depend on the state
+* For `requested` it would be token.
+* For `pending` it would be a serialisation of the `IssuedCertificate`.
+* For `issued` it would be the `IssuedCertificate`.
+
+Encoding the state in the item name, rather than the value makes it clearer when viewing the Secret what state each end entity is in.
+
+**TODO:** inferring the EE state machine state from this and the consumption Secret
+
+#### Trust Secrets
+
+A separate Secret will be used for tracking the state of root CA certificates, for each of the cluster and clients CAs. 
+As described in the description of the process, this Secret can be updated by multiple processes. 
+For example, when the User Operator renews an end-entity certificate and the root CA is not trusted it updates this secret with the root CA certificate, and the CO reacts by ensuring that brokers trust that root CA certificate.
+As such **securing access to this `Secret` is of paramount importance** to ensure the security of the cluster.
+
+```yaml
+kind: Secret
+metdata:
+  name: ${cluster}-${ca-type}-trust # e.g. foo-cluster-ca-trust
+type: strimzi.io/trust
+data:
+  ${fingerprint}.${state}: <PEM encoded Secret>
+```
+
+The `fingerprint` is the fingerprint/thumbprint of the certificate. I.e. the SHA-1 hash of the ASN.1 DER-encoded form of the X509 certificate. The `state` can be any of the states of the CA certificate trust state machine described above.
```

Comment:
two concerns ... first are we sure all characters valid in a certificate fingerprint are valid for a key in the `data` section of a `Secret`? the second one is about how I can get use of this information just watching at the data ... I mean how can I recognize what's the CA cert just seeing a fingerprint and not a Subject or more human readable info like this? Or do you think that because it's all automated we don't really need human readibility of this?